design/outlier-detection-design.md

```diff
@@ -0,0 +1,216 @@
+
+```go
+// OutlierDetection defines the configuration for outlier detection on a service.
+type OutlierDetection struct {
+    // ConsecutiveServerErrors defines The number of consecutive server-side error responses before a consecutive 5xx ejection occurs.
+    // When the backend host encounters consecutive
+    // errors greater than or equal to ConsecutiveServerErrors, it will be
+    // ejected from the load balancing pool.
+    // for HTTP services, a 5xx counts as an error and for TCP services
+    // connection failures and connection timeouts count as an error.
+    // It can be disabled by setting the value to 0.
+    // Defaults to 5.
+    // +optional
+	ConsecutiveServerErrors *uint32 `json:"consecutiveServerErrors,omitempty"`
+    
+    // Interval is the interval at which host status is evaluated.
+    // Defaults to 10s.
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(((\d*(\.\d*)?h)|(\d*(\.\d*)?m)|(\d*(\.\d*)?s)|(\d*(\.\d*)?ms))+)$`
+    Interval *string `json:"interval,omitempty"`
+    
+    // BaseEjectionTime is the base time that a host is ejected for.
+    // A host will remain ejected for a period of time equal to the
+    // product of the ejection base duration and the number of times the host has been ejected.
+    // Defaults to 30s.
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(((\d*(\.\d*)?h)|(\d*(\.\d*)?m)|(\d*(\.\d*)?s)|(\d*(\.\d*)?ms))+)$`
+    BaseEjectionTime *string `json:"baseEjectionTime,omitempty"`
+    
+    // MaxEjectionTime is the maximum time a host will be ejected for.
+    // After this amount of time, a host will be returned to normal operation.
+    // If not specified, the default value (300s) or BaseEjectionTime value is applied, whatever is larger.
+    // Defaults to 300s.
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(((\d*(\.\d*)?h)|(\d*(\.\d*)?m)|(\d*(\.\d*)?s)|(\d*(\.\d*)?ms))+)$`
+    MaxEjectionTime *string `json:"maxEjectionTime,omitempty"`
+    
+    // SplitExternalLocalOriginErrors defines whether to split the local origin errors from the external origin errors.
+    // Defaults to false.
+    // +optional
+    // +kubebuilder:default=false
+    SplitExternalLocalOriginErrors bool `json:"splitExternalLocalOriginErrors"`
+    
+    // ConsecutiveLocalOriginFailure defines the number of consecutive local origin failures before a consecutive local origin ejection occurs.
+    // Parameters take effect only when SplitExternalLocalOriginErrors is true.
+    // Defaults to 5.
+    ConsecutiveLocalOriginFailure *uint32 `json:"consecutiveLocalOriginFailure,omitempty"`
+    
+    // MaxEjectionPercent is the max percentage of hosts in the load balancing pool for the upstream service that can be ejected.
+    // But will eject at least one host regardless of the value here.
+    // Defaults to 10%.
+    // +optional
+    // +kubebuilder:validation:Maximum=100
+    MaxEjectionPercent *uint32 `json:"maxEjectionPercent,omitempty"`
+}
```

Comment:
would be nice to have the `max_ejection_time_jitter` setting here to prevent the host from getting bombarded with connections when it's added back to the healthy pool.

Reply:
this is a good idea